File: R-package/tests/testthat/test-featured-space.R
test_that("FeaturedSpace can be created from features", {
    skip_if_not_installed("maxentcpp")

    n   <- 100L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")

    # Sample indices (0-based): last 10 points
    idx <- as.integer(90:99)

    fs <- maxent_featured_space(n, idx, list(f))
    expect_true(!is.null(fs))
})

test_that("Initial distribution is uniform (all lambdas = 0)", {
    skip_if_not_installed("maxentcpp")

    n   <- 100L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")
    idx <- as.integer(90:99)

    fs <- maxent_featured_space(n, idx, list(f))
    w  <- maxent_model_weights(fs)

    expect_equal(length(w), n)
    expect_equal(sum(w), 1.0, tolerance = 1e-12)
    # Each weight should be 1/n for uniform distribution
    expect_equal(w[1], 1.0 / n, tolerance = 1e-9)
    expect_equal(w[50], 1.0 / n, tolerance = 1e-9)
})

test_that("Model weights sum to ~1.0", {
    skip_if_not_installed("maxentcpp")

    n   <- 50L
    env1 <- runif(n)
    env2 <- runif(n)
    f1   <- maxent_linear_feature(env1, "v1")
    f2   <- maxent_linear_feature(env2, "v2")
    idx  <- as.integer(40:49)

    fs <- maxent_featured_space(n, idx, list(f1, f2))

    # Even before training, weights should sum to 1
    w <- maxent_model_weights(fs)
    expect_equal(sum(w), 1.0, tolerance = 1e-12)
})

test_that("Entropy is non-negative and bounded by log(n)", {
    skip_if_not_installed("maxentcpp")

    n   <- 100L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")
    idx <- as.integer(90:99)

    fs <- maxent_featured_space(n, idx, list(f))
    H  <- maxent_model_entropy(fs)

    expect_true(H >= 0)
    expect_true(H <= log(n) + 1e-9)

    # Uniform distribution has max entropy = log(n)
    expect_equal(H, log(n), tolerance = 1e-9)
})

test_that("Training reduces loss", {
    skip_if_not_installed("maxentcpp")

    set.seed(42)
    n    <- 100L
    env  <- seq(0, 1, length.out = n)
    f    <- maxent_linear_feature(env, "env1")
    idx  <- as.integer(90:99)

    # Record initial loss (all lambdas = 0)
    fs0 <- maxent_featured_space(n, idx, list(f))
    # Set sample expectations to get meaningful loss
    maxent_set_sample_expectations(fs0)
    initial_loss <- maxent_model_loss(fs0)

    # Train a fresh model
    f2  <- maxent_linear_feature(env, "env1")
    fs  <- maxent_featured_space(n, idx, list(f2))
    res <- maxent_fit(fs, max_iter = 300L, convergence = 1e-5)

    expect_true(res$loss <= initial_loss + 1e-9)
    expect_true(res$iterations > 0L)
    expect_equal(length(res$lambdas), 1L)
    expect_true(res$entropy >= 0)
    expect_true(is.logical(res$converged))
})

test_that("Loss is finite after training", {
    skip_if_not_installed("maxentcpp")

    n   <- 80L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")
    idx <- as.integer(70:79)

    fs  <- maxent_featured_space(n, idx, list(f))
    res <- maxent_fit(fs, max_iter = 100L)

    expect_true(is.finite(res$loss))
    expect_true(is.finite(res$entropy))
})

test_that("Prediction returns a numeric vector", {
    skip_if_not_installed("maxentcpp")

    n   <- 50L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")
    idx <- as.integer(40:49)

    fs <- maxent_featured_space(n, idx, list(f))
    maxent_fit(fs, max_iter = 100L)

    # Predict on 5 new points (feature values already evaluated)
    newdata <- matrix(seq(0.1, 0.9, length.out = 5), ncol = 1)
    preds   <- maxent_predict_model(fs, newdata)

    expect_equal(length(preds), 5L)
    expect_true(all(is.finite(preds)))
    expect_true(all(preds >= 0))
})

test_that("Lambda save/load round-trip preserves values", {
    skip_if_not_installed("maxentcpp")

    n    <- 80L
    env1 <- seq(0, 1, length.out = n)
    env2 <- rev(env1)
    f1   <- maxent_linear_feature(env1, "v1")
    f2   <- maxent_linear_feature(env2, "v2")
    idx  <- as.integer(70:79)

    # Train
    fs   <- maxent_featured_space(n, idx, list(f1, f2))
    res1 <- maxent_fit(fs, max_iter = 200L)

    # Save
    tmpfile <- tempfile(fileext = ".lambdas")
    on.exit(unlink(tmpfile))
    maxent_save_lambdas(fs, tmpfile)

    expect_true(file.exists(tmpfile))

    # Load into a fresh space
    f1b <- maxent_linear_feature(env1, "v1")
    f2b <- maxent_linear_feature(env2, "v2")
    fs2 <- maxent_featured_space(n, idx, list(f1b, f2b))
    maxent_load_lambdas(fs2, tmpfile)

    # Weights should match
    w1 <- maxent_model_weights(fs)
    w2 <- maxent_model_weights(fs2)
    expect_equal(w1, w2, tolerance = 1e-6)

    # Density normalizer should match
    info1 <- maxent_space_info(fs)
    info2 <- maxent_space_info(fs2)
    expect_equal(info1$density_normalizer, info2$density_normalizer,
                 tolerance = 1e-6)
})

test_that("maxent_space_info returns correct structure", {
    skip_if_not_installed("maxentcpp")

    n   <- 60L
    env <- seq(0, 1, length.out = n)
    f   <- maxent_linear_feature(env, "env1")
    idx <- as.integer(50:59)

    fs   <- maxent_featured_space(n, idx, list(f))
    info <- maxent_space_info(fs)

    expect_equal(info$num_points,   n)
    expect_equal(info$num_samples,  10L)
    expect_equal(info$num_features, 1L)
    expect_true(is.numeric(info$density_normalizer))
    expect_true(is.numeric(info$linear_predictor_normalizer))
})

test_that("Multi-feature training works", {
    skip_if_not_installed("maxentcpp")

    set.seed(123)
    n    <- 100L
    env1 <- seq(0, 1, length.out = n)
    env2 <- rev(env1)
    idx  <- as.integer(80:99)

    f1 <- maxent_linear_feature(env1, "v1")
    f2 <- maxent_linear_feature(env2, "v2")
    fs <- maxent_featured_space(n, idx, list(f1, f2))

    res <- maxent_fit(fs, max_iter = 300L)

    expect_equal(length(res$lambdas), 2L)
    expect_true(is.finite(res$loss))

    w <- maxent_model_weights(fs)
    expect_equal(sum(w), 1.0, tolerance = 1e-12)
})

File: cpp/tests/test_featured_space.cpp
#include <iostream>
#include <cassert>
#include <cmath>
#include <vector>
#include <memory>
#include <numeric>
#include <sstream>
#include <fstream>
#include <cstdio>
#include <unistd.h>
#include "maxent/feature.hpp"
#include "maxent/featured_space.hpp"

// Helper: compare doubles within tolerance
static bool near(double a, double b, double tol = 1e-9) {
    return std::fabs(a - b) < tol;
}

// Build a small synthetic dataset
// 100 background points, 10 samples, 2 linear features
static std::vector<std::shared_ptr<maxent::Feature>> make_features(
        const std::vector<double>& env1,
        const std::vector<double>& env2)
{
    using namespace maxent;
    auto v1 = std::make_shared<std::vector<double>>(env1);
    auto v2 = std::make_shared<std::vector<double>>(env2);

    double min1 = *std::min_element(env1.begin(), env1.end());
    double max1 = *std::max_element(env1.begin(), env1.end());
    double min2 = *std::min_element(env2.begin(), env2.end());
    double max2 = *std::max_element(env2.begin(), env2.end());

    std::vector<std::shared_ptr<Feature>> features;
    features.push_back(std::make_shared<LinearFeature>(v1, "env1_linear", min1, max1));
    features.push_back(std::make_shared<LinearFeature>(v2, "env2_linear", min2, max2));
    return features;
}

int main() {
    using namespace maxent;

    std::cout << "Testing FeaturedSpace..." << std::endl;

    // ----------------------------------------------------------------
    // Synthetic data: 100 background points, 10 samples
    // ----------------------------------------------------------------
    const int N = 100;
    const int S = 10;

    std::vector<double> env1(N), env2(N);
    for (int i = 0; i < N; ++i) {
        env1[i] = static_cast<double>(i) / (N - 1);  // 0 to 1
        env2[i] = 1.0 - env1[i];                      // 1 to 0
    }

    // Samples at the high end of env1 (indices 90..99)
    std::vector<int> sample_idx(S);
    for (int k = 0; k < S; ++k) sample_idx[k] = 90 + k;

    // ----------------------------------------------------------------
    // Test 1: Construction and initial uniform distribution
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);

        // All lambdas are 0 → linear_predictor is all 0 → density uniform
        auto weights = fs.get_weights();
        assert(static_cast<int>(weights.size()) == N);

        double wsum = 0.0;
        for (double w : weights) wsum += w;
        assert(near(wsum, 1.0, 1e-12));

        // Each weight should be 1/N
        for (double w : weights) {
            assert(near(w, 1.0 / N, 1e-9));
        }

        // Entropy should be log(N)
        double H = fs.get_entropy();
        assert(H >= 0.0);
        assert(near(H, std::log(N), 1e-9));

        std::cout << "  Test 1 (initial uniform distribution): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 2: set_linear_predictor and set_density consistency
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        // Manually set a non-zero lambda
        features[0]->set_lambda(1.0);

        FeaturedSpace fs(N, sample_idx, features);
        // After construction, linear predictor should be sum lambda_j * feature_j

        // Check density sums to density_normalizer
        double dsum = 0.0;
        for (int i = 0; i < N; ++i) dsum += fs.get_density(i);
        assert(near(dsum, fs.get_density_normalizer(), 1e-9));

        // Weights should sum to 1
        auto w = fs.get_weights();
        double wsum = 0.0;
        for (double wi : w) wsum += wi;
        assert(near(wsum, 1.0, 1e-9));

        std::cout << "  Test 2 (linear predictor / density consistency): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 3: increase_lambda updates density correctly
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);

        double loss_before = fs.get_loss();

        // Increase lambda of feature 0
        fs.set_sample_expectations(1.0);
        fs.increase_lambda(0, 0.5);

        // Density should still sum to density_normalizer
        double dsum = 0.0;
        for (int i = 0; i < N; ++i) dsum += fs.get_density(i);
        assert(near(dsum, fs.get_density_normalizer(), 1e-6));

        // Weights should sum to 1
        auto w = fs.get_weights();
        double wsum = 0.0;
        for (double wi : w) wsum += wi;
        assert(near(wsum, 1.0, 1e-9));

        std::cout << "  Test 3 (increase_lambda updates density): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 4: get_weights sums to ~1.0
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        features[0]->set_lambda(2.0);
        features[1]->set_lambda(-0.5);
        FeaturedSpace fs(N, sample_idx, features);

        auto w = fs.get_weights();
        double wsum = std::accumulate(w.begin(), w.end(), 0.0);
        assert(near(wsum, 1.0, 1e-9));

        std::cout << "  Test 4 (get_weights sums to 1): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 5: get_entropy is non-negative and bounded by log(N)
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        features[0]->set_lambda(2.0);
        FeaturedSpace fs(N, sample_idx, features);

        double H = fs.get_entropy();
        assert(H >= 0.0);
        assert(H <= std::log(N) + 1e-9);

        std::cout << "  Test 5 (entropy non-negative and bounded): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 6: Training decreases loss
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);

        // Record initial loss (with all lambdas = 0, sample expectations set)
        fs.set_sample_expectations(1.0);
        double initial_loss = fs.get_loss();

        // Reset and train
        auto features2 = make_features(env1, env2);
        FeaturedSpace fs2(N, sample_idx, features2);
        auto result = fs2.train(200, 1e-5, 1.0);

        assert(result.loss <= initial_loss + 1e-9);
        assert(result.iterations > 0);
        assert(result.lambdas.size() == 2);

        // Weights should still sum to 1
        auto w = fs2.get_weights();
        double wsum = std::accumulate(w.begin(), w.end(), 0.0);
        assert(near(wsum, 1.0, 1e-9));

        // Entropy should be non-negative
        assert(result.entropy >= 0.0);

        std::cout << "  Test 6 (training decreases loss): OK" << std::endl;
        std::cout << "    Initial loss: " << initial_loss
                  << " Final loss: " << result.loss
                  << " Iterations: " << result.iterations << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 7: get_loss returns finite value
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);
        fs.set_sample_expectations(1.0);

        double loss = fs.get_loss();
        assert(std::isfinite(loss));

        std::cout << "  Test 7 (loss is finite): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 8: write_lambdas / read_lambdas round-trip
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);
        auto result = fs.train(100, 1e-5, 1.0);

        // Write to a temp file
        std::string tmpfile = "/tmp/test_maxent_lambdas_" + std::to_string(getpid()) + ".csv";
        fs.write_lambdas(tmpfile);

        // Read into a fresh FeaturedSpace
        auto features2 = make_features(env1, env2);
        FeaturedSpace fs2(N, sample_idx, features2);
        fs2.read_lambdas(tmpfile);

        // Lambdas should match
        for (int j = 0; j < 2; ++j) {
            assert(near(features[j]->lambda(), features2[j]->lambda(), 1e-12));
        }

        // Density normalizer should be close
        assert(near(fs.get_density_normalizer(), fs2.get_density_normalizer(), 1e-6));

        std::cout << "  Test 8 (write/read lambdas round-trip): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 9: batch increase_lambda
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);
        fs.set_sample_expectations(1.0);

        std::vector<double> alphas = {0.1, -0.1};
        fs.increase_lambda(alphas);

        auto w = fs.get_weights();
        double wsum = std::accumulate(w.begin(), w.end(), 0.0);
        assert(near(wsum, 1.0, 1e-9));

        std::cout << "  Test 9 (batch increase_lambda): OK" << std::endl;
    }

    // ----------------------------------------------------------------
    // Test 10: predict() output range
    // ----------------------------------------------------------------
    {
        auto features = make_features(env1, env2);
        FeaturedSpace fs(N, sample_idx, features);
        fs.train(100, 1e-5, 1.0);

        // Predict on a single new point with feature values [0.5, 0.5]
        std::vector<std::vector<double>> newdata = {{0.5, 0.5}};
        auto preds = fs.predict(newdata);

        assert(preds.size() == 1);
        assert(std::isfinite(preds[0]));
        assert(preds[0] >= 0.0);

        std::cout << "  Test 10 (predict output range): OK" << std::endl;
    }

    std::cout << "All FeaturedSpace tests passed!" << std::endl;
    return 0;
}
